mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -1355,4 +1359,28 @@ object ALS extends DefaultParamsReadable[ALS] with Logging {
    * satisfies this requirement, we simply use a type alias here.
    */
   private[recommendation] type ALSPartitioner = org.apache.spark.HashPartitioner
+
+  /**
+   * Private function to clean up all of the shuffles files from the dependencies and their parents.
+   */
+  private[spark] def cleanShuffleDependencies[T](sc: SparkContext, deps: Seq[Dependency[_]],
+      blocking: Boolean = false): Unit = {
+    // If there is no reference tracking we skip clean up.
+    sc.cleaner.foreach{ cleaner =>
+      /**
+       * Clean the shuffles & all of its parents.
+       */
+      def cleanEagerly(dep: Dependency[_]): Unit = {
+        if (dep.isInstanceOf[ShuffleDependency[_, _, _]]) {
+          val shuffleId = dep.asInstanceOf[ShuffleDependency[_, _, _]].shuffleId
+          cleaner.doCleanupShuffle(shuffleId, blocking)
```

Comment:
Does it recursively clean all ancestors, including cached RDDs like `userInBlocks`? Did you test the implementation with many iterations? If the shuffle files are deleted, how does ALS recover from node failures? Please run a job and kill some nodes during the middle and see what happens. My recommendation is cleaning shuffle files up to the cached RDDs and stop there.